File: LeetCode Problems/301 to 400/322. Coin Change/322. Coin Change.cpp
class Solution {
public:
    int coinChange(vector<int> &coins, int amount) {
        int n = coins.size();
        vector<vector<int>> dp(n + 1, vector<int>(amount + 1));
        for (int i = 0; i <= n; i++) {
            for (int j = 0; j <= amount; j++) {
                if (i == 0)
                    dp[i][j] = 10e5;
                else if (j == 0)
                    dp[i][j] = 0;
                else if (coins[i - 1] > j)
                    dp[i][j] = dp[i - 1][j];
                else
                    dp[i][j] = min(dp[i - 1][j], 1 + dp[i][j - coins[i - 1]]);
            }
        }
        return dp[n][amount] == 10e5 ? -1 : dp[n][amount];
    }
};

File: LeetCode Problems/301 to 400/322. Coin Change/322. Coin Change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int dp[] = new int[amount + 1];
        for (int i = 0; i < amount + 1; i++) {
            dp[i] = amount + 1;
        }
        dp[0] = 0;
        int a;
        for (a = 1; a < amount + 1; a++) {
            for (int c : coins) {
                if (a - c >= 0) {
                    dp[a] = Math.min(dp[a], 1 + dp[a - c]);
                }
            }
        }
        if (dp[amount] != amount + 1) {
            return dp[amount];
        } else {
            return -1;
        }
    }
}
